oktapam/client/secrets.go

```diff
@@ -0,0 +1,318 @@
+	if secretFolderResp, foundSecretFolder, secretFolderErr := resolveSecretFolder(ctx, sdkClient, resourceGroupID, projectID, path); secretFolderErr != nil {
+		return nil, secretFolderErr
+	} else if !foundSecretFolder {
+		// resolve_secret requires that you know whether it is a secret or folder, which we do not know here.  if it doesn't resolve as a secret folder, we try again as a secret
+		if secretResp, foundSecret, secretErr := resolveSecret(ctx, sdkClient, resourceGroupID, projectID, path); secretErr != nil {
+			return nil, secretErr
+		} else if !foundSecret {
+			return nil, fmt.Errorf("could not resolve secret or folder with resource group: %s, project: %s, path %s", resourceGroupID, projectID, path)
+		} else {
+			return secretResp, nil
+		}
+	} else {
+		return secretFolderResp, nil
+	}
+}
+
+func ListSecrets(ctx context.Context, sdkClient SDKClientWrapper, resourceGroupID string, projectID string, path string, revealSecrets bool) ([]*wrappers.SecretWrapper, error) {
+	resp, err := resolveSecretOrFolder(ctx, sdkClient, resourceGroupID, projectID, path)
+	if err != nil {
+		return nil, err
+	}
+
+	if *resp.Type == pam.SecretType_FOLDER {
+		parentFolderID := resp.Id
+		parentFolderPath := make([]pam.SecretPath, len(resp.Path)+1)
+		copy(parentFolderPath, resp.Path)
+		parentFolderPath[len(parentFolderPath)-1] = *pam.NewSecretPath(resp.Id, resp.Name)
+
+		return ListSecretsUnderParentFolder(ctx, sdkClient, resourceGroupID, projectID, parentFolderID, parentFolderPath, revealSecrets)
+	}
+
+	if revealSecrets {
+		if secret, err := RevealSecret(ctx, sdkClient, resourceGroupID, projectID, resp.Id); err != nil {
+			return nil, err
+		} else {
+			return []*wrappers.SecretWrapper{secret}, nil
+		}
+	}
+	return []*wrappers.SecretWrapper{ConvertResolveSecretOrFolderResponseToSecretNoContents(resourceGroupID, projectID, resp)}, nil
+}
+
+func ListSecretsUnderParentFolder(ctx context.Context, sdkClient SDKClientWrapper, resourceGroupID string, projectID string, parentFolderID string, parentFolderPath []pam.SecretPath, revealSecrets bool) ([]*wrappers.SecretWrapper, error) {
+	var list []*wrappers.SecretWrapper
+	request := sdkClient.SDKClient.SecretsAPI.ListSecretFolderItems(ctx, sdkClient.Team, resourceGroupID, projectID, parentFolderID)
+
+	for {
+		resp, httpResp, callErr := request.Execute()
+		if httpResp != nil {
+			if _, err := checkStatusCodeFromSDK(httpResp, 200); err != nil {
+				return nil, err
+			}
+		} else if callErr != nil {
+			return nil, callErr
+		}
+
+		for _, s := range resp.List {
+			if s.Type != pam.SecretType_FOLDER {
+				if revealSecrets {
+					if secret, err := RevealSecret(ctx, sdkClient, resourceGroupID, projectID, s.Id); err != nil {
+						return nil, err
+					} else {
+						list = append(list, secret)
+					}
+				} else {
+					list = append(list, ConvertSecretOrFolderListResponseToSecret(parentFolderPath, s))
+				}
+			}
+		}
+
+		linkHeader := httpResp.Header.Get("Link")
```

Comment:
Should all this go in a helper / utils place?  I know terraform likes copy/pasting code, but i imagine this could be shared with other places?

Reply:
I think the last time this came up, we said that each of these implementations is slightly different in ways that make it a bit cumbersome to have a generic function.  

Additionally, my preference would be to have the SDK handle this instead of the provider or any other code that uses it.